polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -3434,8 +3164,137 @@ mlir::Location MLIRASTConsumer::getMLIRLocation(clang::SourceLocation loc) {
   return FileLineColLoc::get(ctx, fileId, lineNumber, colNumber);
 }
 
+llvm::GlobalValue::LinkageTypes
+MLIRASTConsumer::getLLVMLinkageType(const clang::FunctionDecl &FD,
+                                    bool shouldEmit) {
+  if (!FD.hasBody() || !shouldEmit)
+    return llvm::GlobalValue::LinkageTypes::ExternalLinkage;
+  else if (auto CC = dyn_cast<CXXConstructorDecl>(&FD))
+    return CGM.getFunctionLinkage(GlobalDecl(CC, CXXCtorType::Ctor_Complete));
+  else if (auto CC = dyn_cast<CXXDestructorDecl>(&FD))
+    return CGM.getFunctionLinkage(GlobalDecl(CC, CXXDtorType::Dtor_Complete));
+
+  return CGM.getFunctionLinkage(&FD);
+}
+
+mlir::LLVM::Linkage
+MLIRASTConsumer::getMLIRLinkage(llvm::GlobalValue::LinkageTypes LV) {
+  switch (LV) {
+  case llvm::GlobalValue::LinkageTypes::InternalLinkage:
+    return LLVM::Linkage::Internal;
+  case llvm::GlobalValue::LinkageTypes::ExternalLinkage:
+    return LLVM::Linkage::External;
+  case llvm::GlobalValue::LinkageTypes::AvailableExternallyLinkage:
+    return LLVM::Linkage::AvailableExternally;
+  case llvm::GlobalValue::LinkageTypes::LinkOnceAnyLinkage:
+    return LLVM::Linkage::Linkonce;
+  case llvm::GlobalValue::LinkageTypes::WeakAnyLinkage:
+    return LLVM::Linkage::Weak;
+  case llvm::GlobalValue::LinkageTypes::WeakODRLinkage:
+    return LLVM::Linkage::WeakODR;
+  case llvm::GlobalValue::LinkageTypes::CommonLinkage:
+    return LLVM::Linkage::Common;
+  case llvm::GlobalValue::LinkageTypes::AppendingLinkage:
+    return LLVM::Linkage::Appending;
+  case llvm::GlobalValue::LinkageTypes::ExternalWeakLinkage:
+    return LLVM::Linkage::ExternWeak;
+  case llvm::GlobalValue::LinkageTypes::LinkOnceODRLinkage:
+    return LLVM::Linkage::LinkonceODR;
+  case llvm::GlobalValue::LinkageTypes::PrivateLinkage:
+    return LLVM::Linkage::Private;
+  }
+
+  llvm_unreachable("Unexpected linkage");
+}
+
+void MLIRASTConsumer::setMLIRFunctionVisibility(
+    mlir::FunctionOpInterface &function, const clang::FunctionDecl &FD,
+    bool shouldEmit) {
+  llvm::GlobalValue::LinkageTypes LV = getLLVMLinkageType(FD, shouldEmit);
+  SymbolTable::Visibility visibility =
+      (LV == llvm::GlobalValue::InternalLinkage ||
+       LV == llvm::GlobalValue::PrivateLinkage || !FD.isDefined() ||
+       FD.hasAttr<CUDAGlobalAttr>() || FD.hasAttr<CUDADeviceAttr>() ||
+       !shouldEmit)
+          ? SymbolTable::Visibility::Private
+          : SymbolTable::Visibility::Public;
+  SymbolTable::setSymbolVisibility(function, visibility);
```

Comment:
Maybe we can do some sorting here to avoid calling `getLLVMLinkage` or other functions if `shouldEmit` is false.